apps/cms.c

```diff
@@ -885,10 +892,22 @@ int cms_main(int argc, char **argv)
             goto end;
     }
 
-    in = bio_open_default(infile, 'r',
-                          binary_files ? FORMAT_BINARY : informat);
-    if (in == NULL)
-        goto end;
+    if (digesthex != NULL) {
+        if (infile != NULL
+            || operation != SMIME_SIGN
+            || !(flags & CMS_DETACHED)) {
+            BIO_printf(bio_err, "Invalid use of -digest\n");
+            goto end;
+        }
+        digestbin = OPENSSL_hexstr2buf(digesthex, &digestlen);
+        if (digestbin == NULL)
+            goto end;
```

Comment:
Give an error message stating that the hex input was invalid.

Reply:
`OPENSSL_hexstr2buf` already raises errors such as `ERR_raise(ERR_LIB_CRYPTO, CRYPTO_R_ILLEGAL_HEX_DIGIT);` whenever it returns NULL, so I think an error message is already printed. Is that not enough?

Reply:
Those low-level errors like
```
80B297A8407F0000:error:07800066:common libcrypto routines:hexstr2buf_sep:illegal hex digit:crypto/o_str.c:157:
```
are pretty hard to read and do not give sufficient context.
Saying that the `-digest` argument is not a valid hex string is much more user-friendly.